### DISCUSSION TITLE
Does `fs.utimes` invoke `utime()` or `utimes()`?

### DISCUSSION THREAD

**Discussion by jsumners-nr** [nodejs/help#4610] [Q&A]
I have a test against `fs.utimes` that is flaky. The test:

1. creates a temporary file
2. invokes `fs.utimes` against that file to set specific accessed & modified times
3. checks that the times are expected in the callback invoked by `fs.utimes` via `fs.statSync`

This test is flaking out by the accessed time occasionally, being set to an unexpected time (creation time of the temp file?), e.g. `1732540522.827 == 1732539600`.

This comment https://github.com/nodejs/node/issues/30212#issuecomment-549021998 is suggesting that, ultimately, the passed in times are being converted to nanosecond precision. But if I look at:

https://github.com/nodejs/node/blob/b17a1fb1cec919c09aa3401c040b79e35c5ed82e/src/node_file.cc#L2934-L2965

it looks like Node.js will be invoking `utime()` instead of `utimes()` (at least, I think that `static void UTimes` function is what underlies `fs.utimes`). According to https://linux.die.net/man/2/utimes, only `utimes()` supports nanosecond precision.

Is my finding correct? Or am I missing something?

**Comment by juanarbol** [comment]
No, is using `futimens` in Linux

I'll try to build the backtrace for you.

uv_fs_utime calls an `INIT(FUTIME);`
https://github.com/libuv/libuv/blob/c6d43bea09093545e0f199921b2d33548a2a1f85/src/unix/fs.c#L1931

Which ends on Linux in a `futimens`
https://github.com/libuv/libuv/blob/c6d43bea09093545e0f199921b2d33548a2a1f85/src/unix/fs.c#L236

~and on macOS `utimes`
https://github.com/libuv/libuv/blob/c6d43bea09093545e0f199921b2d33548a2a1f85/src/unix/fs.c#L1164~ Nevermind

**Reply by jsumners-nr** [reply]
So, regardless of platform, there is room for conversion error. However, the test I was investigating is _likely_ only failing on macOS -- https://github.com/newrelic/node-newrelic/pull/2790

**Reply by juanarbol** [reply]
macOS uses `futimes`, each platform seems to have different "storage" precision/resolution. I don't know if we can do anything as every OS seems to behave different.
